llpc/lower/llpcSpirvLowerGlobal.cpp

```diff
@@ -2419,6 +2377,15 @@ void SpirvLowerGlobal::lowerBufferBlock() {
               assert(getElemPtr->getNumIndices() >= 2);
               SmallVector<Value *, 8> indices;
 
+              // Check if Global Variable type is the same as GEP source type. For opaque pointers these types can be
+              // different because opaque pointers are removing zero-index GEP instructions. Without opaque pointers
+              // LLVM-IR would contain additional zero-index GEP which will be between "getElemPtr" and "global". These
+              // two GEP instructions would be at some point merged to single one.
+              if (getElemPtr->getSourceElementType() != global.getValueType()) {
+                appendZeroIndexToMatchTypes(m_builder, indices, getElemPtr->getSourceElementType(),
+                                            global.getValueType());
+              }
+
```

Comment:
Is this really necessary? The indices are really only used to create a new GEP, and the new GEP doesn't need trailing zero indices with opaque pointers.

The only case to watch out for is if the block indices themselves are missing, but that case could be handled by not inserting a new GEP at all -- right?

Reply:
You covered all cases why I did it, but this change is for leading zeros, not for trailing.

Type and indices of GEP are different for opaque pointer and since Block Index is equal to GEP's indices[1], I had to add these leading zeros. Adding leading zeros also solved issue with new GEP which was based on Global Variable Type.

For new GEP I could used existing indices (without adding leading zeros) and change Type of the GEP while creating, but not sure if I can assume that Block Index == 0 if Types are different for oldGEP and GV ?

Reply:
Ah, thank you for the explanation.

If the oldGEP base type is different from the GV type, then yes, I think you can assume that the block index is 0 and form the newGEP with the same base type as oldGEP.